### PR TITLE
bernoulli_ only taking in float, includes dropout

### PR DESCRIPTION
**Pull request by samdow**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #523
* #521

### CODE REVIEW

**Comment by vfdev-5** [comment]
@samdow checked per_sample_grad script for torchvision classification models (https://github.com/pytorch/functorch/issues/445#issuecomment-1043254485) using your branch and checks are passing for models without BN.
Do you know what's blocking to merge your stack ?

**Comment by samdow** [comment]
> @samdow checked per_sample_grad script for torchvision classification models (https://github.com/pytorch/functorch/issues/445#issuecomment-1043254485) using your branch and checks are passing for models without BN.
Do you know what's blocking to merge your stack ?

Hi @vfdev-5 sorry I missed this comment earlier. So so glad to hear that this is working! That gives me a lot more faith in this. The last things I had hoped to do was double check all paths of dropouts (alpha, alpha_feature, inplace of all of those, etc). That's now in this PR, so fingers crossed that this passes tests like it does locally and then I'll merge this today.

Still not done on dropout are removing the test failures because that comes from `freeze_rng`. I'm planning to do that in a separate PR just so this one didn't get too messy

File: test/test_eager_transforms.py [hunk 1]
```diff
@@ -2319,6 +2320,71 @@ def slice_weights(batched_weights, index):
...
+                loss = loss_fn(output, targets)
+                return loss
+
+            grad_weights, loss = grad_and_value(compute_loss)(weights, batch, targets)
+            new_weights = []
+            with torch.no_grad():
+                for grad_weight, weight in zip(grad_weights, weights):
+                    new_weights.append(weight - grad_weight * lr)
+            # NB: return looks weird because torch.vmap must return Tensors
+            return (loss, *new_weights)
```
Comment (zou3519) [inline test/test_eager_transforms.py:2365]: nit: Wait, does it still have the limitation that vmap must return Tensors? I thought we relaxed vmap to be able to accept functions that return pytrees of Tensors

File: test/test_eager_transforms.py [hunk 2]
```diff
@@ -2319,6 +2320,71 @@ def slice_weights(batched_weights, index):
         self.assertEqual(result_loss, expected_loss)
         self.assertEqual(result_weights, expected_weights)
 
+    @parametrize("dropout_layer", [nn.Dropout, nn.AlphaDropout, nn.FeatureAlphaDropout])
+    def test_find_learning_rate_ensembling(self, device, dropout_layer):
+        # This example mimics what a user might do when trying to find the optimal learning rate. They would
+        # want to run a bunch of models with the same behavior (including the same dropout!) and have them
+        # each run with different learning rates. Specifically, this is an example of using same randomness with vmap
```
Comment (zou3519) [inline test/test_eager_transforms.py:2327]: this is a cool application!

File: test/test_vmap.py [hunk 3]
```diff
@@ -3547,6 +3546,38 @@ def test_randperm(self, device, randomness, use_generator):
             for i in range(B0):
                 assert torch.allclose(vmap_result[i], expected)
 
+    @parametrize('randomness', ['error', 'same', 'different'])
+    def test_feature_dropout(self, device, randomness):
+        # special case because functional feature dropout expects a batch dimension, so it doesn't match the
+        # pattern for other randomness
```
Comment (zou3519) [inline test/test_vmap.py:3552]: should this just be `test_dropout` ? Or by "feature_dropout" do you mean `feature_alpha_dropout`?

File: test/test_vmap.py [hunk 4]
```diff
@@ -3547,6 +3546,38 @@ def test_randperm(self, device, randomness, use_generator):
...
+        supported_ops = [
+            lambda t: torch.nn.functional.dropout(torch.ones_like(t), training=True),
+            lambda t: torch.nn.functional.alpha_dropout(torch.ones_like(t), training=True),
+            lambda t: torch.nn.functional.feature_alpha_dropout(t, training=True),
+            lambda t: torch.nn.functional.dropout2d(t, training=True),
+            lambda t: torch.nn.functional.dropout3d(t.unsqueeze(-1), training=True),
+        ]
+
+        for op in supported_ops:
+            passed = torch.ones([4, 3, 3, 3, 3])
```
Comment (zou3519) [inline test/test_vmap.py:3564]: Does this need `device=device`?

Comment (zou3519) [reply test/test_vmap.py:3564]: Also, prefer to use the `B0` convention here so we know which dimension is being vmappd over
```
B0 = ...
passed = torch.ones(B0, 3, 3, 3, 3])
```

File: test/test_vmap.py [hunk 5]
```diff
@@ -3547,6 +3546,38 @@ def test_randperm(self, device, randomness, use_generator):
...
+            vmap_result = vmap(op, randomness=randomness)(passed)
+            torch.manual_seed(seed)
+            if randomness == 'different':
+                expected = op(passed.flatten(0, 1))
+                expected = expected.reshape(passed.shape[0], passed.shape[1], *expected.shape[1:])
+                assert torch.allclose(vmap_result, expected)
+            else:
+                expected = op(passed[0])
+                for i in range(4):
+                    assert torch.allclose(vmap_result[i], expected)
```
Comment (zou3519) [inline test/test_vmap.py:3579]: nit: torch.allclose -> self.assertEqual

**Review by zou3519 — COMMENTED** [review]
cool!
